### PR TITLE
Fix incorrect projection analysis when aggregation expression contains monotonic functions

### LINKED ISSUE #52075: LOGICAL_ERROR: Arguments of 'lcm' have incorrect data types
Found by SQLancer -> https://s3.amazonaws.com/clickhouse-test-reports/51692/42a7ec77895c3b3def77192ab7ecb8b3e7692beb/sqlancer__release_.html

Reproducer:

```
CREATE TABLE t0
(
    `c0` Int32
)
ENGINE = MergeTree
PARTITION BY -c0
ORDER BY tuple()
SETTINGS allow_suspicious_indices = 1;

INSERT INTO t0(c0) VALUES (-283770613);

SELECT abs(log(radians(0)))
FROM t0
GROUP BY lcm(-sign(t0.c0), -t0.c0);
```

Reproduces in:
* master KO
* 23.6 KO
* 23.3 KO
* 23.2 OK
* 23.1 OK
* 22.12 OK
* 22.8 OK

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix incorrect projection analysis when aggregation expression contains monotonic functions. This fixes https://github.com/ClickHouse/ClickHouse/issues/52075

This PR should be merged after #52152  and it will re-enable `optimize_use_implicit_projections` by default.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 85e757d9a99699ba816f749849ec69298789bdff with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52151/85e757d9a99699ba816f749849ec69298789bdff/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Processors/QueryPlan/Optimizations/optimizeUseAggregateProjection.cpp [hunk 1]
```diff
@@ -287,7 +287,7 @@ ActionsDAGPtr analyzeAggregateProjection(
 {
     auto proj_index = buildDAGIndex(*info.before_aggregation);
 
-    MatchedTrees::Matches matches = matchTrees(*info.before_aggregation, *query.dag);
+    MatchedTrees::Matches matches = matchTrees(*info.before_aggregation, *query.dag, false /* check_monotonicity */);
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/Optimizations/optimizeUseAggregateProjection.cpp:290]: I do not understand why exactly monotonicity is the issue here.
We use this function for read-in-order analysis, and monotonicity is needed.

Comment (amosbird) [reply src/Processors/QueryPlan/Optimizations/optimizeUseAggregateProjection.cpp:290]: Answered in Telegram.

**Comment by KochetovNicolai** [comment]
#52160

**Comment by KochetovNicolai** [comment]
#49400

**Review by KochetovNicolai — APPROVED** [review]

**Comment by amosbird** [comment]
@KochetovNicolai I'll move that test case to `01739_index_hint.sql` instead, which is disabled for test analyzer.

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**